keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,990 @@
+ClusterNetworkPolicy API and the ClusterDefaultNetworkPolicy API to complement
+the developer focused NetworkPolicy API in Kubernetes.
+
+## Motivation
+
+Kubernetes provides NetworkPolicy resources to control traffic within a
+cluster. NetworkPolicies focus on expressing a developers intent to secure
+their applications. Thus, in order to satisfy the needs of a security admin,
+we propose to introduce new set of APIs that capture the administrators intent.
+
+### Goals
+
+The goals for this KEP are to satisfy two key user stories:
+1. As a cluster administrator, I want to enforce irrevocable guardrails that
+   all workloads must adhere to in order to guarantee the safety of my clusters.
+2. As a cluster administrator, I want to deploy a default set of policies to all
+   workloads that may be overridden by the developers if needed.
+
+There are several unique properties that we need to add in order accomplish the
+user stories above.
+1. Deny rules and, therefore, hierarchical enforcement of policy
+2. Semantics for a cluster-scoped policy object that may include
+   namespaces/workloads that have not been created yet.
+3. Backwards compatibility with existing Kubernetes Network Policy API
+
+### Non-Goals
+
+Our mission is to solve the most common use cases that cluster admins have.
+That is, we don't want to solve for every possible policy permutation a user
+can think of. Instead, we want to design an API that addresses 90-95% use cases
+while keeping the mental model easy to understand and use.
+
+Additionally, this proposal is squarely focused on solving the needs of the
+Cluster Administrator. It is not intended to solve:
+1. Logging / error reporting for Network Policy
+2. Kubernetes Node policies
+3. New policy selectors (services, service accounts, etc.)
+4. Tools/CLI to debug/explain NetworkPolicies, cluster-scoped policies and their
+   impact on workloads.
+
+## Proposal
+
+In order to achieve the two primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following two new resources
+under `networking.k8s.io` API group:
+- ClusterNetworkPolicy
+- ClusterDefaultNetworkPolicy
+
+### ClusterNetworkPolicy resource
+
+A ClusterNetworkPolicy resource will help the administrators set strict
+security rules for the cluster, i.e. a developer CANNOT override these rules
+by creating NetworkPolicies that applies to the same workloads as the
+ClusterNetworkPolicy does.
+
+Unlike the NetworkPolicy resource in which each rule represents an allowed
+traffic, ClusterNetworkPolicy will enable administrators to set `Authorize`,
+`Deny` or `Allow` as the action of each rule. ClusterNetworkPolicy rules should
+be read as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the ClusterNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+In terms of precedence, the aggregated `Authorize` rules (all ClusterNetworkPolicy
```

Comment:
Sure. I will follow up and change it to `AlwaysAllow`.

Reply:
I'm still... uneasy about overloading a single "action" field to include a concept of precedence as well as the pure action.

I guess let's see what the broader SIG / API reviewers have to say about this.

Reply:
Would it make more sense to simply go to `int` priority values so "exceptions" were less exceptional?

Reply:
I second revisiting int priority values so that the order of precedence is absolutely clear without various combinations of ambiguous keywords like Authorize, Except, AlwaysAllow and so on

Reply:
Recap of sig-net discussion:

I'm not hard-against any particular model.  What I want to balance is:

* simplicity (less is more)
* expressiveness (can I make it do what I need)
* familiarity (is it like something else I know)

If numeric priorities win big points on familiarity, then I am willing to overlook a bit more complexity. 

The Allow/Deny model with `exceptions` feels simple to me, but I ACK that maybe just me.